main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1145 @@
+/*
+ * Copyright 2015 Magnus Madsen, Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.KindedAst.Expr
+import ca.uwaterloo.flix.language.ast.Type.getFlixType
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, KindedAst, LevelEnv, Name, RigidityEnv, Scheme, SemanticOp, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.constraintgeneration.{RestrictableChooseConstraintGeneration, SchemaConstraintGeneration}
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+import ca.uwaterloo.flix.util.collection.MapOps
+
+import scala.collection.mutable.ListBuffer
+
+object ConstraintGeneration {
+
+  def run(root: KindedAst.Root)(implicit flix: Flix): Map[Symbol.DefnSym, (List[Constraint], Type, Type, RigidityEnv)] = {
+
+    // Skip this phase unless it is activated
+    if (!flix.options.xtyper) {
+      return Map.empty
+    }
+
+    val result = flix.phase("ConstraintGeneration") {
+      ParOps.mapValues(root.defs) {
+        case defn =>
+          implicit val context: Context = Context.empty()
+          implicit val r: KindedAst.Root = root
+          val (tpe, eff) = visitExp(defn.exp)
+          val constrs = context.constrs.toList
+          val renv = context.renv
+          (constrs, tpe, eff, renv)
+      }
+    }
+
+    // report the constraints if directed
+    if (flix.options.xprintconstraints) {
+      result.foreach {
+        case (sym, (tconstrs, tpe, eff, renv)) =>
+          println(sym)
+          println("Type: " + tpe)
+          println("Effect: " + eff)
+          println("Rigid: " + renv.s)
+          println("Type constraints: " + tconstrs)
+          println()
+      }
+    }
+
+    // return the result
+    result
+  }
+
+  sealed class Constraint
+
+  object Constraint {
+    case class Equality(tpe1: Type, tpe2: Type, lenv: LevelEnv, loc: SourceLocation) extends Constraint
```

Comment:
I think my most important comment is that we need two Equality constraints: 

- One where types must be equals, and
- One where we expect one type.

So we can keep generating useful error messages.

Reply:
I was thinking that we could keep it with two constraints and then add a field `provenance` or whatever that tells where the constraint came from. That way the constraint reduction doesn't need to care about it.